classes/manager.php

```diff
@@ -144,5 +144,81 @@ public static function is_factorid_valid($factorname, $factorid, $user) {
 
         return false;
     }
+
+    /**
+     * Function to display to the user that they cannot login, then log them out.
+     *
+     * @return void
+     */
+    public static function cannot_login() {
+        self::mfa_logout();
+        print_error('error:notenoughfactors', 'tool_mfa', new \moodle_url('/'));
+    }
+
+    /**
+     * Logout user.
+     *
+     * @return void
+     */
+    public static function mfa_logout() {
+        $authsequence = get_enabled_auth_plugins();
+        foreach ($authsequence as $authname) {
```

Comment:
What is this for?

I can see prelogout_hook and prelogout_hook but no logoutpage_hook

more dead code from before?

Reply:
This hook exists in a couple of auth plugins e.g. mnet, shibboleth etc just found from a grep. I think it might be useful to keep, but this code I didnt write, just moved over from lib.php

Reply:
ah yes I missed this as it's called from the logout.php